### ISSUE TITLE
FP16 training error

### ISSUE THREAD

**Issue by gxzks** [OpenNMT/OpenNMT-py#1645]
Here is my script for training:
`CUDA_VISIBLE_DEVICES=0,1,2,3,4,5,6,7 onmt_train -data onmt_data/tiny -save_model onmt_save/tiny \
     -layers 6 -rnn_size 512 -word_vec_size 512 -transformer_ff 2048 -heads 8 \
     -encoder_type transformer -decoder_type transformer -position_encoding \
     -train_steps 100000 -max_generator_batches 2 -dropout 0.1 \
     -batch_size 4096 -batch_type tokens -normalization tokens -accum_count 1 \
     -optim adam -adam_beta2 0.998 -decay_method noam -warmup_steps 8000 -learning_rate 2 \
     -max_grad_norm 0 -param_init 0 -param_init_glorot \
     -label_smoothing 0.1 -valid_steps 10000 \
     -save_checkpoint_steps 10000 -keep_checkpoint 1 \
     -world_size 8 -gpu_ranks 0 1 2 3 4 5 6 7 \
     -log_file tiny.log \
     -report_every 100 \
     -model_dtype fp16 -apex_opt_level O1`

I used `model_dtype=fp16` and `apex_opt_level=O1` for training.

Here is the training log:
![image](https://user-images.githubusercontent.com/46742345/69028694-b96cf680-0a0d-11ea-93d1-a2a422a413e3.png)

xent went to be nan and finally training crashed.
![image](https://user-images.githubusercontent.com/46742345/69028757-f20cd000-0a0d-11ea-8773-568951057da6.png)
 
When I used `apex_opt_level=O2`, training went well but the result was bad compared to fp32 training, about 3 bleu lower.

**Comment by vince62s** [comment]
Read this: https://github.com/NVIDIA/apex/issues/475
I fixed it making sure we use the legacy FusedAdam
So try using Fusedadam as the optimizer instead of Adam. It work fine and it is faster.

**Comment by gxzks** [comment]
I got the error when I use fusedadam:
![image](https://user-images.githubusercontent.com/46742345/69038078-1673a700-0a24-11ea-8e15-c1b4ce6e9542.png)
Is the opennmt version wrong?
I use pip to install and my version is `OpenNMT-py==1.0.0rc2`

**Comment by vince62s** [comment]
this is more an apex issue. when did you install / recompile it ?

**Comment by gxzks** [comment]
I just reinstall apex with 
`pip install -v --no-cache-dir --global-option="--cpp_ext" --global-option="--cuda_ext" ./`
 this error still occured

**Comment by gxzks** [comment]
I reinstall apex with an extra flag `--global-option="--deprecated_fused_adam"`
And it works.

**Comment by gaussmao** [comment]
I have a similar issue as mentioned above, and I've submitted it on the OpenNMT forum (https://forum.opennmt.net/t/about-the-issues-with-opennmt-py-in-machine-translation-models-really-looking-forward-to-the-experts-response-and-assistance-many-thanks/5634). Do you have time to help clarify it for me? Thank you very much!
